File: plugin/src/main/java/org/jukeboxmc/plugin/perms/model/GroupImpl.java
package org.jukeboxmc.plugin.perms.model;

import java.util.HashSet;
import java.util.Set;
import lombok.Getter;
import lombok.RequiredArgsConstructor;
import lombok.Setter;
import lombok.experimental.Accessors;
import org.jukeboxmc.plugin.perms.api.model.Group;

/**
 * @author Kaooot
 * @version 1.0
 */
@RequiredArgsConstructor
@Accessors(fluent = true)
public class GroupImpl implements Group {

    @Getter
    private final String id;
    @Getter
    private final int priority;
    @Getter
    private final Set<String> inheritedPermissions = new HashSet<>();

    @Setter
    @Getter
    private Set<String> permissions = new HashSet<>();
    @Setter
    @Getter
    private String nameTag;
    @Setter
    @Getter
    private String chatFormat;
    @Setter
    @Getter
    private String parentId;

    public static GroupImpl ofApiModel(Group apiModel) {
        final GroupImpl group = new GroupImpl(apiModel.id(), apiModel.priority());
        group.nameTag(apiModel.nameTag());
        group.chatFormat(apiModel.chatFormat());
        group.permissions(apiModel.permissions());
        group.parentId(apiModel.parentId());

        return group;
    }
}

File: plugin/src/main/kotlin/org/jukeboxmc/plugin/perms/model/GroupImpl.kt
package org.jukeboxmc.plugin.perms.model

import org.jukeboxmc.plugin.perms.api.model.Group

/**
 * @author Kaooot
 * @version 1.0
 */
class GroupImpl(private val id: String, private val priority: Int) : Group {

    private val inheritedPermissions: MutableSet<String> = HashSet()
    private var permissions: MutableSet<String> = HashSet()
    private var nameTag: String = ""
    private var chatFormat: String = ""
    private var parentId: String = ""

    companion object {
        fun ofApiModel(apiModel: Group): GroupImpl {
            val group = GroupImpl(apiModel.id(), apiModel.priority())
            group.nameTag = apiModel.nameTag()
            group.chatFormat = apiModel.chatFormat()
            group.permissions = apiModel.permissions()
            group.parentId = apiModel.parentId()

            return group
        }
    }

    override fun id() = this.id
    override fun priority() = this.priority
    override fun permissions() = this.permissions
    override fun nameTag() = this.nameTag
    override fun chatFormat() = this.chatFormat
    override fun parentId() = this.parentId

    fun inheritedPermissions() = this.inheritedPermissions

    fun permissions(permissions: MutableSet<String>) {
        this.permissions = permissions
    }

    fun nameTag(nameTag: String) {
        this.nameTag = nameTag
    }

    fun chatFormat(chatFormat: String) {
        this.chatFormat = chatFormat
    }

    fun parentId(parentId: String) {
        this.parentId = parentId
    }
}